packages/sdk/src/cosmwasmclient.ts

```diff
@@ -0,0 +1,228 @@
+  MsgExecuteContract,
+  MsgInstantiateContract,
+  MsgStoreCode,
+  StdFee,
+  StdSignature,
+} from "./types";
+
+const defaultUploadFee: StdFee = {
+  amount: [
+    {
+      amount: "5000",
+      denom: "ucosm",
+    },
+  ],
+  gas: "1000000", // one million
+};
+
+const defaultInitFee: StdFee = {
+  amount: [
+    {
+      amount: "5000",
+      denom: "ucosm",
+    },
+  ],
+  gas: "500000", // 500k
+};
+
+const defaultExecFee: StdFee = {
+  amount: [
+    {
+      amount: "5000",
+      denom: "ucosm",
+    },
+  ],
+  gas: "200000", // 200k
+};
+
+export interface SigningCallback {
+  (signBytes: Uint8Array): Promise<StdSignature>;
+}
+
+interface SigningData {
+  readonly senderAddress: string;
+  readonly signCallback: SigningCallback;
+}
+
+export interface GetNonceResult {
+  readonly accountNumber: number;
+  readonly sequence: number;
+}
+
+export interface PostTxResult {
+  readonly logs: readonly Log[];
+  readonly rawLog: string;
+  readonly transactionHash: string;
+}
+
+export class CosmWasmClient {
+  public static makeReadOnly(url: string): CosmWasmClient {
+    return new CosmWasmClient(url);
+  }
+
+  public static makeWritable(
+    url: string,
+    senderAddress: string,
+    signCallback: SigningCallback,
+  ): CosmWasmClient {
+    return new CosmWasmClient(url, {
+      senderAddress: senderAddress,
+      signCallback: signCallback,
+    });
+  }
+
+  private readonly restClient: RestClient;
+  private readonly signingData: SigningData | undefined;
+
+  private get senderAddress(): string {
+    if (!this.signingData) throw new Error("Signing data not set in this client");
+    return this.signingData.senderAddress;
+  }
+
+  private get signCallback(): SigningCallback {
+    if (!this.signingData) throw new Error("Signing data not set in this client");
+    return this.signingData.signCallback;
+  }
+
+  private constructor(url: string, signingData?: SigningData) {
+    this.restClient = new RestClient(url);
+    this.signingData = signingData;
+  }
+
+  public async chainId(): Promise<string> {
+    const response = await this.restClient.nodeInfo();
+    return response.node_info.network;
+  }
+
+  /**
+   * Returns account number and sequence.
+   *
+   * @param address returns data for this address. When unset, the client's sender adddress is used.
+   */
+  public async getNonce(address?: string): Promise<GetNonceResult> {
+    const account = (await this.restClient.authAccounts(address || this.senderAddress)).result.value;
+    return {
+      accountNumber: account.account_number,
+      sequence: account.sequence,
+    };
+  }
+
+  public async postTx(tx: Uint8Array): Promise<PostTxResult> {
+    const result = await this.restClient.postTx(tx);
+    if (result.code) {
+      throw new Error(`Error when posting tx. Code: ${result.code}; Raw log: ${result.raw_log}`);
+    }
+    return {
+      logs: parseLogs(result.logs) || [],
+      rawLog: result.raw_log || "",
+      transactionHash: result.txhash,
```

Comment:
this is a hex string, I believe, how about parsing into Uint8Array?

Reply:
Uppercase hex is the primary printable transaction ID in the Tendermint ecosystem, which is why I kept it a string. Maybe we can model that better by renaming `transactionHash` -> `transactionId`.

Reply:
4c7124d12c9da060654ae1ab118451b705a8e745 adds format checks and format guarantees in the API. I think this is more convenient than exposing binary data.

Reply:
Fair enough